src/hitchhiker/tree/core.cljc

```diff
@@ -275,43 +297,62 @@ throwable error."
 (defn index-node-keys
   "Calculates the separating keys given the children of an index node"
   [children]
-  (mapv last-key (butlast children)))
+  (into []
+        (map last-key)
+        (pop children)))
 
-(declare ->IndexNode)
+(declare index-node)
 
-(defrecord IndexNode [children storage-addr op-buf cfg]
+(defrecord IndexNode [children storage-addr op-buf cfg
+                      *last-key-cache]
   IResolve
   (index? [this] true)
   (dirty? [this] (not (poll! storage-addr)))
+  (dirty! [this]
+    (assoc this
+           :storage-addr (promise-chan)
+           :*last-key-cache (volatile! ::nothing)))
   (resolve [this] (go-try this))
   (last-key [this]
-    ;;TODO should optimize by caching to reduce IOps (can use monad)
-    (last-key (peek children)))
+    (<-cache *last-key-cache
```

Comment:
Nice. I guess this is the reason for most of the speedups? The benchmarks test write operations mostly, but we should see large speedups on reads by this as well.

Reply:
Yes it contributes quite a bit.

Reply:
Good, very nice.